File: 74. Search a 2D Matrix.js
/**
 * @param {number[][]} matrix
 * @param {number} target
 * @return {boolean}
 */
var searchMatrix = function(matrix, target) {
    if( matrix == null || matrix.length == 0 ) {
        return false;
    }

    for ( var arr of matrix ) {
        var result = binarySearch( arr, target );
        if ( result ) {
            return true;
        }
    }
    return false;

    function binarySearch( arr, target ) {
        if( arr == null || arr.length == 0 ) {
            return false;
        }

        var mid;
        var start = 0;
        var end = arr.length -1;

        while ( start + 1 < end ) {
            mid = start + Math.round(( end - start ) / 2);
            if ( arr[mid] == target ) {
                return true;
            } else if( arr[mid] > target ) {
                end = mid;
            } else {
                start = mid;
            }
        }

        if ( arr[start] == target || arr[end] == target ) {
            return true;
        }

        return false;
    }
};

File: 74. Search a 2D Matrix.py
"""
假设是一个3*4的矩阵，low = 0, high = 11, 那么mid = 5
5 = 4x + 1
"""
class Solution(object):
    def searchMatrix(self, matrix, target):
        if not matrix or len(matrix) == 0 or len(
                matrix[0]) == 0 or target == None:
            return False

        m, n = len(matrix), len(matrix[0])
        low, high = 0, m * n - 1

        while low + 1 < high:
            mid = (low + high) / 2
            x, y = mid / n, mid % n
            if matrix[x][y] < target:
                low = mid
            else:
                high = mid

        x, y = low / n, low % n
        if matrix[x][y] == target:
            return True

        x, y = high / n, high % n
        if matrix[x][y] == target:
            return True

        return False
